File: agents/_lib/tools.ts
/**
 * RAG 工具定义 —— OpenAI Agents SDK 版本。
 *
 * 所有工具从 ./loader.ts 读取 agents/_data 下的本地 RAG 数据，
 * 通过 Agents SDK 的 `tool()` 工厂声明，供 Agent 的 agent loop 调用。
 */

import { tool } from "@openai/agents";
import { z } from "zod";
import {
	getDocumentMeta,
	getDocumentStructure,
	getPageContent,
	listDocuments,
} from "./loader.ts";

// ── RAG 工具 1: 获取文档结构，供 LLM 推理定位 ──────────────────────────────
const searchDocument = tool({
	name: "searchDocument",
	description:
		"搜索知识库文档，获取文档元信息和树状索引结构，用于推理定位相关页码。" +
		"当用户询问知识库相关内容时，优先调用此工具。",
	parameters: z.object({
		query: z.string().describe("用户的搜索问题"),
		docId: z
			.string()
			.nullable()
			.describe("文档ID，传 null 则自动选择第一个可用文档"),
	}),
	async execute({ query, docId }) {
		// 若未指定 docId，自动选第一个可用文档
		if (!docId) {
			const docs = await listDocuments();
			if (docs.length === 0) {
				return {
					error:
						"知识库为空，请先运行 python3 python-script/prepare_rag_data.py 生成文档",
				};
			}
			docId = docs[0].docId;
		}

		const meta = await getDocumentMeta(docId);
		if (!meta) {
			return { error: `文档 ${docId} 不存在，请检查文档 ID` };
		}

		const structure = await getDocumentStructure(docId);

		return {
			docId,
			query,
			meta,
			structure,
			instruction:
				"请仔细分析以上文档结构（structure），找出与用户问题最相关的章节，" +
				"确定对应的页码范围（start_index 到 end_index），" +
				"然后调用 fetchPages 工具获取该页码范围的原文内容，再基于原文回答用户问题。",
		};
	},
});

// ── RAG 工具 2: LLM 推理定位后精准拉取原文 ────────────────────────────────
const fetchPages = tool({
	name: "fetchPages",
	description:
		"根据页码范围获取文档原文内容。在 searchDocument 定位到相关章节后调用此工具获取原文。",
	parameters: z.object({
		docId: z.string().describe("文档ID"),
		pages: z
			.string()
			.describe(
				"页码范围，支持格式: '5' / '5-7' / '5-7,12' / '5-7,12,15-16'，单次最多 20 页",
			),
	}),
	async execute({ docId, pages }) {
		const content = await getPageContent(docId, pages);
		if (content.length === 0) {
			return { error: "未找到指定页面，请检查页码范围是否正确" };
		}

		// 获取文档名称
		const meta = await getDocumentMeta(docId);
		const docName = meta?.doc_name ?? docId;

		const totalChars = content.reduce((sum, p) => sum + p.content.length, 0);

		return {
			type: "citation_pages",
			docId,
			docName,
			pages,
			pageCount: content.length,
			totalChars,
			content: content.map((p) => ({
				page: p.page,
				content: p.content,
				preview: p.content.slice(0, 400),
			})),
		};
	},
});

// ── 原有辅助工具 ────────────────────────────────────────────────────────────
const getWeather = tool({
	name: "getWeather",
	description: "Get the current weather for a city",
	parameters: z.object({
		city: z.string().describe("City name"),
	}),
	async execute({ city }) {
		const conditions = ["sunny", "cloudy", "rainy"];
		const temp = Math.floor(Math.random() * 30) + 5;
		return {
			city,
			temperature: temp,
			condition: conditions[Math.floor(Math.random() * conditions.length)],
		};
	},
});

const getUserTimezone = tool({
	name: "getUserTimezone",
	description:
		"Get the server's current timezone and local time.",
	parameters: z.object({}),
	async execute() {
		return {
			timezone: Intl.DateTimeFormat().resolvedOptions().timeZone,
			localTime: new Date().toLocaleTimeString(),
		};
	},
});

const calculate = tool({
	name: "calculate",
	description: "Perform a math calculation with two numbers.",
	parameters: z.object({
		a: z.number().describe("First number"),
		b: z.number().describe("Second number"),
		operator: z
			.enum(["+", "-", "*", "/", "%"])
			.describe("Arithmetic operator"),
	}),
	async execute({ a, b, operator }) {
		const ops: Record<string, (x: number, y: number) => number> = {
			"+": (x, y) => x + y,
			"-": (x, y) => x - y,
			"*": (x, y) => x * y,
			"/": (x, y) => x / y,
			"%": (x, y) => x % y,
		};
		if (operator === "/" && b === 0) {
			return { error: "Division by zero" };
		}
		return {
			expression: `${a} ${operator} ${b}`,
			result: ops[operator](a, b),
		};
	},
});

/** 返回 RAG 场景下的全部工具（给 Agent 构造时传入 tools 字段）。 */
export function buildRagTools() {
	return [searchDocument, fetchPages, getWeather, getUserTimezone, calculate];
}

File: agents/_tools.py
"""RAG tool definitions for OpenAI Agents SDK (Python).

5 tools: searchDocument, fetchPages, getWeather, getUserTimezone, calculate
"""

import json
import time
import random
from typing import Annotated

from agents import function_tool

from ._loader import (
    list_documents,
    get_document_meta,
    get_document_structure,
    get_page_content,
    parse_page_range,
)
from ._logger import create_logger

logger = create_logger("tools")


@function_tool
def search_document(
    query: Annotated[str, "User's search query"],
    doc_id: Annotated[str, "Document ID. Pass empty string to auto-select first available document."] = "",
) -> str:
    """Search knowledge base documents. Returns document metadata and structure index for locating relevant pages.
    Always call this first when answering knowledge-base questions."""

    logger.log(f"searchDocument called: query=\"{query}\", doc_id=\"{doc_id}\"")

    # Auto-select first available document if no doc_id specified
    if not doc_id:
        docs = list_documents()
        logger.log(f"listDocuments returned {len(docs)} docs")
        if not docs:
            return json.dumps({
                "error": "Knowledge base is empty. Please run prepare_rag_data.py first.",
            }, ensure_ascii=False)
        doc_id = docs[0]["docId"]

    meta = get_document_meta(doc_id)
    if not meta:
        return json.dumps({
            "error": f"Document '{doc_id}' not found.",
        }, ensure_ascii=False)

    structure = get_document_structure(doc_id)

    # When no tree-structure index exists, instruct to fetch all pages for small docs
    if not structure:
        page_count = meta.get("page_count", 0)
        suggested_pages = f"1-{page_count}" if page_count <= 20 else "1-20"
        return json.dumps({
            "docId": doc_id,
            "query": query,
            "meta": meta,
            "structure": None,
            "instruction": (
                f"This document has no tree-structure index but contains {page_count} pages. "
                f"Please call fetchPages with pages=\"{suggested_pages}\" to retrieve the content, "
                "then answer the user's question based on the retrieved text."
            ),
        }, ensure_ascii=False)

    return json.dumps({
        "docId": doc_id,
        "query": query,
        "meta": meta,
        "structure": structure,
        "instruction": (
            "Analyze the document structure above, find sections most relevant to the user's question, "
            "determine the page range (start_index to end_index), "
            "then call fetchPages to retrieve the original text and answer based on it."
        ),
    }, ensure_ascii=False)


@function_tool
def fetch_pages(
    doc_id: Annotated[str, "Document ID"],
    pages: Annotated[str, "Page range, e.g. '5-7,12,15-16'. Max 20 pages per call."],
) -> str:
    """Fetch specific pages from a document. Use after searchDocument to retrieve actual content."""

    logger.log(f"fetchPages called: doc_id=\"{doc_id}\", pages=\"{pages}\"")

    page_list = parse_page_range(pages)
    if not page_list:
        return json.dumps({"error": "Invalid page range"}, ensure_ascii=False)

    content = get_page_content(doc_id, page_list)
    if not content:
        return json.dumps({"error": "No pages found"}, ensure_ascii=False)

    # Get document name for citation
    meta = get_document_meta(doc_id)
    doc_name = meta.get("doc_name", doc_id) if meta else doc_id

    total_chars = sum(len(p["content"]) for p in content)

    return json.dumps({
        "type": "citation_pages",
        "docId": doc_id,
        "docName": doc_name,
        "pages": pages,
        "pageCount": len(content),
        "totalChars": total_chars,
        "content": [
            {"page": p["page"], "content": p["content"], "preview": p["content"][:400]}
            for p in content
        ],
    }, ensure_ascii=False)


@function_tool
def get_weather(city: Annotated[str, "City name"]) -> str:
    """Get current weather for a city (mock data)."""
    conditions = ["sunny", "cloudy", "rainy"]
    temp = random.randint(5, 35)
    return json.dumps({
        "city": city,
        "temperature": temp,
        "condition": random.choice(conditions),
    })


@function_tool
def get_user_timezone() -> str:
    """Get the server's current timezone and local time."""
    return json.dumps({
        "timezone": time.tzname[0] if time.tzname[0] else "UTC",
        "localTime": time.strftime("%Y-%m-%d %H:%M:%S"),
    })


@function_tool
def calculate(
    a: Annotated[float, "First number"],
    b: Annotated[float, "Second number"],
    operator: Annotated[str, "Arithmetic operator: +, -, *, /, %"],
) -> str:
    """Perform a math calculation with two numbers."""
    ops = {
        "+": lambda x, y: x + y,
        "-": lambda x, y: x - y,
        "*": lambda x, y: x * y,
        "/": lambda x, y: x / y if y != 0 else None,
        "%": lambda x, y: x % y if y != 0 else None,
    }
    if operator not in ops:
        return json.dumps({"error": f"Unknown operator: {operator}"})
    result = ops[operator](a, b)
    if result is None:
        return json.dumps({"error": "Division by zero"})
    return json.dumps({
        "expression": f"{a} {operator} {b}",
        "result": result,
    })


# Tool collection for RAG agent
RAG_TOOLS = [search_document, fetch_pages, get_weather, get_user_timezone, calculate]
